### PR TITLE
expression: enable SHA2 function pushdown to TiKV

### LINKED ISSUE #67087: expression: enable SHA2 function pushdown to TiKV coprocessor
### Feature Request

Enable the `SHA2` function to be pushed down to TiKV for coprocessor evaluation.

### Background

TiKV already implements SHA2 in its coprocessor (`components/tidb_query_expr/src/impl_encryption.rs`), supporting all hash lengths (SHA-0/224/256/384/512). The TiDB side also has:
- PbCode mapping: `tipb.ScalarFuncSig_SHA2` (in `builtin_encryption.go`)
- distsql deserialization support (in `distsql_builtin.go`)

However, `SHA2` is missing from the `scalarExprSupportedByTiKV` allowlist in `infer_pushdown.go`, so it is never pushed down to TiKV. `MD5` and `SHA1` are already in the list.

### Motivation

Pushing SHA2 down to TiKV reduces data transfer and allows TiKV to evaluate SHA2 in WHERE clauses locally, improving query performance for SHA2-based filtering.

Related: https://github.com/pingcap/tidb-tools/pull/886

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: close #67087

Problem Summary: The `SHA2` function is not pushed down to TiKV for coprocessor evaluation, even though TiKV already implements it. This causes unnecessary data transfer when SHA2 is used in WHERE clauses.

Related: https://github.com/pingcap/tidb-tools/pull/886 (from customer)

### What changed and how does it work?

Added `ast.SHA2` to the `scalarExprSupportedByTiKV` allowlist in `pkg/expression/infer_pushdown.go`, alongside the already-supported `MD5` and `SHA1`.

All infrastructure was already in place:
- TiKV: `SHA2` implemented in `components/tidb_query_expr/src/impl_encryption.rs` (SHA-0/224/256/384/512)
- tipb: `ScalarFuncSig_SHA2` defined in protobuf
- TiDB: PbCode mapping (`builtin_encryption.go`) and distsql deserialization (`distsql_builtin.go`) already existed

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
  > Verified with tiup playground (PD + TiKV + two TiDB instances: stock vs modified) that:
  > - EXPLAIN shows `Selection` moves from `root` to `cop[tikv]` after the change
  > - All hash lengths (0/224/256/384/512) produce identical results
  > - All column types (varchar, char, text, blob, binary, int, bigint, double, decimal, datetime, date, time, bit, enum, set) produce identical results
  > - NULL handling and invalid hash length (123) produce identical results
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
The `SHA2` function can now be pushed down to TiKV for coprocessor evaluation, improving performance for queries that use `SHA2` in `WHERE` clauses.
```

## Summary by CodeRabbit

* **New Features**
  * SHA2 hashing is now eligible for push-down execution to TiKV, improving query performance for SHA-2–based predicates.

* **Tests**
  * Updated and added integration tests to cover SHA2 query planning and plan_tree expectations for pushed-down execution.

### CODE REVIEW

File: tests/integrationtest/t/planner/core/casetest/pushdown/push_down.test [hunk 1]
```diff
@@ -35,6 +35,7 @@ explain format = 'plan_tree' select json_unquote(a) from t2;
 explain format = 'plan_tree' select i * 2 from t;
 explain format = 'plan_tree' select DATE_FORMAT(t, '%Y-%m-%d %H') as date from t;
 explain format = 'plan_tree' select md5(s) from t;
+explain format = 'plan_tree' select sha2(s, 256) from t;
```
Comment (zanmato1984) [inline tests/integrationtest/t/planner/core/casetest/pushdown/push_down.test]: I think this added case still does not verify the pushdown path enabled by this PR. `select sha2(s, 256) from t` can still produce a root `Projection` over a TiKV scan, so it does not prove that `SHA2` itself is evaluated in `cop[tikv]`. Could we switch this to a case that forces TiKV-side evidence (for example, a `WHERE sha2(...) = ...` predicate that yields `Selection cop[tikv]`), or add a direct pb-conversion regression test instead?

File: tests/integrationtest/r/planner/core/casetest/pushdown/push_down.result [hunk 2]
```diff
@@ -121,6 +121,11 @@ Projection	root		date_format(planner__core__casetest__pushdown__push_down.t.t, %
 explain format = 'plan_tree' select md5(s) from t;
 id	task	access object	operator info
 Projection	root		md5(planner__core__casetest__pushdown__push_down.t.s)->Column
+└─TableReader	root		data:TableFullScan
+  └─TableFullScan	cop[tikv]	table:t	keep order:false, stats:pseudo
+explain format = 'plan_tree' select sha2(s, 256) from t;
+id	task	access object	operator info
+Projection	root		sha2(planner__core__casetest__pushdown__push_down.t.s, 256)->Column
```
Comment (zanmato1984) [inline tests/integrationtest/r/planner/core/casetest/pushdown/push_down.result]: The expected plan here still keeps `sha2(...)` in a root `Projection`, so this golden file would pass even if `SHA2` remained evaluated in TiDB above a TiKV scan. Could we change the case to one that shows TiKV-side evidence instead, such as `Selection cop[tikv]` for a `WHERE sha2(...) = ...` predicate, or `Projection cop[tikv]` if you want to prove projection pushdown directly?

File: tests/integrationtest/t/planner/core/casetest/pushdown/push_down.test (hunk 1 above)
Comment (joechenrh) [reply tests/integrationtest/t/planner/core/casetest/pushdown/push_down.test]: Good catch — you're right, the projection case doesn't prove anything since it stays at root regardless.

Changed to `select * from t where sha2(s, 256) = 'abc'` which now shows `Selection cop[tikv]` with `sha2(...)` inside, confirming TiKV-side evaluation.

Done in https://github.com/pingcap/tidb/commit/5feb1e7cae9e8edae0cf66b4574f51e6fd947f25

File: tests/integrationtest/r/planner/core/casetest/pushdown/push_down.result (hunk 2 above)
Comment (joechenrh) [reply tests/integrationtest/r/planner/core/casetest/pushdown/push_down.result]: Same fix as above — switched to a WHERE predicate so the plan now shows `Selection cop[tikv]` with `sha2(...)`.

Done in https://github.com/pingcap/tidb/commit/5feb1e7cae9e8edae0cf66b4574f51e6fd947f25

File: pkg/expression/infer_pushdown.go [hunk 3]
```diff
@@ -241,7 +241,7 @@ func scalarExprSupportedByTiKV(ctx EvalContext, sf *ScalarFunction) bool {
 		ast.Sysdate, /* ast.StrToDate, */
 
 		// encryption functions.
-		ast.MD5, ast.SHA1, ast.UncompressedLength,
+		ast.MD5, ast.SHA1, ast.SHA2, ast.UncompressedLength,
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/expression/infer_pushdown.go:244]: SHA2 follows the exact same pattern as MD5 and SHA1, which have been in the pushdown allowlist for years without any version gating. There's no version negotiation mechanism in `canFuncBePushed` for any coprocessor function — the codebase relies on proper cluster upgrade ordering and the runtime `expr_pushdown_blacklist` table as a safety valve. Adding version-specific gating just for SHA2 would be inconsistent with every other function in the list.

File: tests/integrationtest/t/planner/core/casetest/pushdown/push_down.test [hunk 4]
```diff
@@ -35,6 +35,7 @@ explain format = 'plan_tree' select json_unquote(a) from t2;
 explain format = 'plan_tree' select i * 2 from t;
 explain format = 'plan_tree' select DATE_FORMAT(t, '%Y-%m-%d %H') as date from t;
 explain format = 'plan_tree' select md5(s) from t;
+explain format = 'plan_tree' select * from t where sha2(s, 256) = 'abc';
```
[the remark this answers is not in the document]
Comment (joechenrh) [reply tests/integrationtest/t/planner/core/casetest/pushdown/push_down.test:38]: The SHA2 function itself already has unit test coverage for different modes and edge cases in the expression package. This PR only enables pushdown — the EXPLAIN test proves TiKV-side evaluation happens (`Selection cop[tikv]` with `sha2(...)` inside). Semantic correctness of SHA2 computation is TiKV's responsibility and is tested on that side.

File: pkg/expression/infer_pushdown.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (joechenrh) [reply pkg/expression/infer_pushdown.go:244]: Fair point, though the "encryption functions" comment predates this PR — it already covered MD5, SHA1, and UncompressedLength. I'd prefer to keep the diff minimal and not mix in a comment rename with this functional change.

**Review by windtalker — APPROVED** [review]
lgtm

**Review by zanmato1984 — APPROVED** [review]
lgtm
